promql/engine.go

```diff
@@ -115,6 +115,12 @@ func (e ErrStorage) Error() string {
 	return e.Err.Error()
 }
 
+// QueryEngine defines the interface for the *promql.Engine, so it can be replaced, wrapped or mocked.
+type QueryEngine interface {
+	NewInstantQuery(ctx context.Context, q storage.Queryable, opts QueryOpts, qs string, ts time.Time) (Query, error)
+	NewRangeQuery(ctx context.Context, q storage.Queryable, opts QueryOpts, qs string, start, end time.Time, interval time.Duration) (Query, error)
```

Comment:
Something completely outside of the scope of the PR, but I wonder why the interface is coupled to the prometheus storage format. This parameter looks like something that should be part of the constructor of a concrete implementation. Not having `storage.Queryable` here would allow us to query different data sources which are not necessarily related to Prometheus.

Reply:
As far as I can see, `storage.Queryable` is just an abstract interface that is not tied to the Prometheus storage format (and that is also used with other storage backends in other systems already). Am I missing something?

Reply:
Yes this is true and flexible to an extent, but we are still limited by the parameters of that interface. For example, there is no good way to request only certain labels for a series since such a parameter does not exist in the `Select` statement. In Thanos we now have a way to run PromQL in a distributed manner, and telling a remote engine to execute a query is not possible through that interface.

Reply:
Everything is possible through context 🙈 (we already do that in Thanos, and it's 💩 )

On serious note, we probably need some good repro example on where is the problem e.g. test/one go file and I think it could make sense to move queryable out of this interface for flexibility (or some other way like Select with some more generic params 🤔 ). Not for this PR for sure